### ISSUE TITLE
How to extract piano-roll matrix representation of a given measure?

### ISSUE THREAD

**Issue by adamoudad** [cuthbertLab/music21#370]
I would like to extract the piano-roll representation of each measure of a midi file.

Pretty_midi library has a function [get_piano_roll](http://craffel.github.io/pretty-midi/#pretty_midi.PrettyMIDI.get_piano_roll) for converting a midi file to piano-roll representation, returning its matrix of pitch activations for all time-steps.

However, I did not find how to divide a midi file into measures, as Music21 does when using [converter.parse](http://web.mit.edu/music21/doc/moduleReference/moduleConverter.html?highlight=converter%20parse#music21.converter.parse)

Is there a way to, from a Measure object, get a numpy ndarray corresponding to its piano-roll representation? That is, the representation which can be effectively plotted using [.plot('pianoroll')](http://web.mit.edu/music21/doc/usersGuide/usersGuide_22_graphing.html)

Thank you in advance,
Thanks for your amazing work.
Adam.

**Comment by mscuthbert** [comment]
Hi Adam -- it's absolutely possible.  Just look through the code used to generate .plot and you can see how to do it.  The user's guide should get you through all the steps.  There's no one line code to do so, and not likely to be one in music21 itself.

**Comment by adamoudad** [comment]
Ok, thank you for the precisions !

**Comment by adamoudad** [comment]
Hello again,
Is the code you are referring  to in https://github.com/cuthbertLab/music21/blob/master/music21/graph/plot.py ?
I cannot find the place where the stream is converted to piano-roll matrix representation, any pointer to a specific code or function will be much appreciated ! :)

**Comment by qbeer** [comment]
This might be late to help but it seems to me that the `pianoroll` plot is using the `HorizontalBarPitchSpaceOffset` class so basically you could extract the data with that

```python
from music21 import corpus, graph

paths = corpus.getComposer('bach')

streamObj = corpus.parse(paths[0])

gr = graph.plot.HorizontalBarPitchSpaceOffset(streamObj)
gr.extractData()

print(gr.data)
```

```bash
[['F2', [(23.0, 0.9375), (47.0, 0.9375), (79.0, 0.9375)], {}],
 ['', [], {}],
 ['', [], {}],
 ['', [], {}],
 ['A', [(4.0, 0.9375), (28.0, 0.9375)], {}],
 ['B♭',
  [(5.0, 0.4375),
   (17.0, 0.9375),
   (29.0, 0.4375),
   (41.0, 0.9375),
   (73.0, 0.4375)],
  {}],
 ['B', [(77.0, 0.9375)], {}],
 ['C3',
  [(5.5, 0.4375),
   (15.0, 0.9375),
   (18.0, 0.9375),
   (21.0, 1.9375),
   (29.5, 0.4375),
   (39.0, 0.9375),
   (42.0, 0.9375),
   (45.0, 1.9375),
   (53.0, 1.9375),
   (59.0, 0.4375),
   (60.5, 0.4375),
   (78.0, 0.9375)],
  {}],
 ['', [], {}],
 ['D',
  [(3.0, 0.9375),
   (6.0, 0.4375),
   (19.0, 0.4375),
   (27.0, 0.9375),
   (30.0, 0.4375),
   (43.0, 0.4375),
   (59.5, 0.4375),
   (71.0, 0.4375),
   (73.5, 0.4375),
   (76.5, 0.4375)],
  {}],
 ['', [], {}],
...
 ['F', [(11.0, 0.9375), (35.0, 0.9375), (71.0, 0.9375)], {}]]
```

On the other hand I am not educated in music, does any of you could help me how `Measure Number` is calculated? I am seeing values ranging from `0-80` in the arrays next to the notes, and I assume that the first value should be used since the second values I reckon are the lengths, but on the plot `Measure Number` is ranging from `0-20ish`.

![image](https://user-images.githubusercontent.com/24634931/77329289-15c68680-6d1e-11ea-8658-55fce5f42eaf.png)

Thanks in advance. :)

Alex

**Comment by pushkarjajoria** [comment]
If you are looking for a pianoroll ndarray from the 'mxl' files of music21, you could use the piano_roll function from [pitchscapes](https://github.com/robert-lieck/pitchscapes/blob/master/pitchscapes/reader.py#L118) library. Just pass the filepath to this. 

```
from pitchscapes.reader import piano_roll
paths = corpus.getComposer('bach')

pr = []
for path in paths:
    pr.append(piano_roll(path))
```

You might need to install is manually using pip install -e

**Comment by adamoudad** [comment]
Sorry for the late answer.

I ended up assuming my MIDI files are 4/4 and using pretty_midi get the pianoroll, although this is not optimal.
 
@qbeer your solution looks ok, but some more processing is needed to get a nice pianoroll matrix. Especiallly, I am not sure if I can rely on the notation for the pitches. Does "F" always mean "F4"?

@pushkarjajoria I did not know about pitchscapes, thanks for sharing!
